src/XrdCrypto/XrdCryptosslX509.cc

```diff
@@ -432,14 +408,26 @@ void XrdCryptosslX509::CertType()
 //_____________________________________________________________________________
 void XrdCryptosslX509::SetPKI(XrdCryptoX509data newpki)
 {
-   // Set PKI
-
-   // Cleanup key first
-   if (pki)
-      delete pki;
-   if (newpki)
-      pki = new XrdCryptosslRSA((EVP_PKEY *)newpki, 1);
-
+   // SetPKI:
+   // if newpki is null does nothing
+   // if newpki contains a consistent private & public key we take ownership
+   //   so that this->PKI()->status will be kComplete.
+   // otherwise, newpki is not consistent:
+   // if the previous PKI() was null or was already kComplete it is and reset
+   //   so that this->PKI()->status will be kInvalid.
+
+   if (!newpki) return;
+
+   std::unique_ptr<XrdCryptoRSA> tmprsa(new XrdCryptosslRSA((EVP_PKEY *)newpki, 1));
```

Comment:
After looking at the changes more thoroughly, I think the changes are all fine. The only suggestion I have is to use `std::make_unique` rather than calling new directly (maybe without using `auto` if you don't like that):
```suggestion
   auto tmprsa = std::make_unique<XrdCryptosslRSA>((EVP_PKEY*)newpki, 1));
```
Note that there is a mismatch of types between `std::unique_ptr<XrdCryptoRSA>` and `new XrdCryptosslRSA` in the code above, which should be fixed in any case.